composable_kernel/include/tensor_operation/reduction_functions_blockwise.hpp

```diff
@@ -23,247 +23,161 @@
  * SOFTWARE.
  *
  *******************************************************************************/
-#ifndef CK_REDUCTION_FUNCTIONS_BLOCKWISE_HPP
-#define CK_REDUCTION_FUNCTIONS_BLOCKWISE_HPP
+#ifndef CK_REDUCTION_FUNCTIONS_PARTITIONED_BLOCKWISE_HPP
+#define CK_REDUCTION_FUNCTIONS_PARTITIONED_BLOCKWISE_HPP
 
 #include "data_type.hpp"
 
 #include "reduction_common.hpp"
 #include "reduction_operator.hpp"
-#include "reduction_functions_binop.hpp"
+#include "reduction_functions_accumulate.hpp"
 
 namespace ck {
 
-template <typename buffer2dDescType,
-          bool blockIsOneRow,
+template <typename buffer1dDescType,
+          typename AccDataType,
+          index_t BlockSize,
+          index_t MThreadClusterSize,
+          index_t KThreadClusterSize,
+          bool reorder_thread_clusters,
           typename opReduce,
-          NanPropagation_t nanPropaOpt>
-struct BlockwiseReduction_2d_block_buffer
+          bool propagate_nan>
+struct PartitionedBlockwiseReduction_1d_block_buffer
 {
-    using compType = typename opReduce::dataType;
+    static constexpr auto buffer1dDesc = buffer1dDescType{};
 
-    static constexpr auto buffer2dDesc = buffer2dDescType{};
+    static_assert(BlockSize == MThreadClusterSize * KThreadClusterSize,
+                  "The product of cluster lengths should be same as BlockSize!");
+    static_assert(KThreadClusterSize > 1, "Parallel reduction need work on at least two elements");
 
-    static constexpr index_t BlockSize =
-        blockIsOneRow ? buffer2dDesc.GetLength(Number<1>{}) : buffer2dDesc.GetLength(Number<0>{});
-    static constexpr index_t NumBlocks =
-        blockIsOneRow ? buffer2dDesc.GetLength(Number<0>{}) : buffer2dDesc.GetLength(Number<1>{});
-    using binop = detail::binop_with_nan_check<nanPropaOpt, opReduce, compType>;
+    static_assert(buffer1dDesc.GetElementSize() == BlockSize,
+                  "The buffer size should be the same as BlockSize!");
+
+    using Accumulation = detail::accumulate_with_nan_check<propagate_nan, opReduce, AccDataType>;
 
-    // This interface does not accumulate on indices
     template <typename BufferType>
-    __device__ static void
-    Reduce(BufferType& block_buffer, index_t toReduceBlocks, compType& accuData)
+    __device__ static void Reduce(BufferType& block_buffer,
+                                  AccDataType& accuData,
+                                  index_t thread_m_cluster_id,
```

Comment:
thread cluster id should be calculated inside this class, by using get_thread_local_1d_id and cluster descriptor

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/22d438ae9e02b674e1656263df07799ee06fb466/composable_kernel/include/tensor_operation/blockwise_tensor_slice_transfer_v4r1.hpp#L78-L79

Reply:
Will change this after the P.R

Reply:
My most recent thought is that the `[thread_m_cluster_id, thread_k_cluster_id]` should be calculated outside `reduction_function_blockwise.hpp` to keep them consistent with that used in `gridwise_2d_reduction_blockwise.hpp`.  And the lay out used to calculate the  `[thread_m_cluster_id, thread_k_cluster_id]` could be M_K or K_M,  even though the LDS buffer descriptor can always be in lay out M_K.